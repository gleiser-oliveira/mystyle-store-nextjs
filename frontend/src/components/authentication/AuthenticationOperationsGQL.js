import gql from 'graphql-tag';
import { perPage } from '../../config';

export const SIGNUP_MUTATION = gql`
    mutation SIGNUP_MUTATION($email: String!, $name: String!, $password: String!) {
        signup(email: $email, name: $name, password: $password) {
            id
            email
            name
        }
    }
`;

export const SIGNIN_MUTATION = gql`
    mutation SIGNIN_MUTATION($email: String!, $password: String!) {
        signin(email: $email, password: $password) {
            id
            email
            name
        }
    }
`;

export const SIGNOUT_MUTATION = gql`
    mutation SIGNOUT_MUTATION {
        signout {
            message
        }
    }
`;

export const CURRENT_USER_QUERY = gql`
    query CURRENT_USER_QUERY { 
        me {
            id
            email
            name
            permissions
            cart {
                id
                quantity
                product {
                    id
                    price
                    title
                    image
                    description
                }
            }
        }
    }
`;

export const REQUEST_PASSWORD_RESET_MUTATION = gql`
    mutation REQUEST_PASSWORD_RESET_MUTATION($email: String!) {
        requestPasswordReset(email: $email) {
            message
        }
    }
`;

export const RESET_PASSWORD_MUTATION = gql`
    mutation RESET_PASSWORD_MUTATION($resetToken: String!, $password: String!, $confirmPassword: String!) {
        resetPassword(resetToken: $resetToken, password: $password, confirmPassword: $confirmPassword) {
            id
            name
            email
        }
    }
`;

export const HAS_PERMISSION_QUERY = gql`
    query { 
        users {
            id
            email
            name
            permissions
        }
    }
`;

export const ALL_USERS_QUERY = gql`
    query ALL_USERS_QUERY($skip: Int = 0, $first: Int = ${perPage}) { 
        users(first: $first, skip: $skip, orderBy: name_ASC) {
            id
            email
            name
            permissions
        }
    }
`;

export const USERS_PAGINATION_QUERY = gql`
    query USERS_PAGINATION_QUERY {
        usersConnection {
            aggregate {
                count
            }
        }
    }
`;

export const ALL_PERMISSIONS_QUERY = gql`
    query { 
        permissions {
          id
          permission
          description
        }
    }
`;

export const UPDATE_USER_PERMISSIONS_MUTATION = gql`
    mutation UPDATE_USER_PERMISSIONS_MUTATION($id: ID! $permissions: [Permission]) {
        updateUserPermissions(id: $id permissions: $permissions) {
            id
            email
            name
            permissions
        }
    }
`;
