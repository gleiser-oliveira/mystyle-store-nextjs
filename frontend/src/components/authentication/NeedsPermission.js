import React from 'react';
import PropTypes from 'prop-types';

const NeedsPermission = props => {
    if (!props.user)
        return null;
    if (props.user.permissions.includes('ADMIN'))
        return props.children;
    const message = props.showMessage ? <div><p>You have no permission to access this resource.</p></div> : null;
    const fragment = ((!props.ownerId) || (props.ownerId === props.user.id)) && props.user.permissions.some(permission => props.neededPermission.includes(permission)) ? props.children : message;
    return fragment;
};

NeedsPermission.propTypes = {
    user: PropTypes.object,
    ownerId: PropTypes.string,
    neededPermission: PropTypes.array,
};

export default NeedsPermission;
