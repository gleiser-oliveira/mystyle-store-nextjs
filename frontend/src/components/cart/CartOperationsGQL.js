import gql from 'graphql-tag';

export const ADD_PRODUCT_TO_CART_MUTATION = gql`
    mutation ADD_PRODUCT_TO_CART_MUTATION($id: ID!){
        addProductToCart(id: $id) {
            id
            quantity
        }
    }
`;

export const REMOVE_CART_PRODUCT_MUTATION = gql`
    mutation REMOVE_CART_PRODUCT_MUTATION($id: ID!) {
        removeCartProduct(id: $id) {
            id
        }
    }
`;

export const REMOVE_PRODUCT_FROM_CART_MUTATION = gql`
    mutation REMOVE_PRODUCT_FROM_CART_MUTATION($id: ID!) {
        removeProductFromCart(id: $id) {
            id
            quantity
        }
    }
`;

export const CLEAR_CART_MUTATION = gql`
    mutation CLEAR_CART_MUTATION {
        removeAllProductsFromUserCart {
            count
        }
    }
`;
