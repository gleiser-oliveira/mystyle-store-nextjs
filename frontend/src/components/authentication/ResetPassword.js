import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Mutation } from 'react-apollo';
import { Form } from '../theme/generalStyles';
import { REQUEST_PASSWORD_RESET_MUTATION, RESET_PASSWORD_MUTATION, CURRENT_USER_QUERY } from './AuthenticationOperationsGQL';
import { ErrorMessage, ConfirmationMessage } from '../message/Message';

class ResetPassword extends Component {
    state = {
        email: this.props.email || '',
        password: '',
        confirmPassword: '',
    };

    saveToState = (e) => {
        this.setState({ [e.target.name]: e.target.value });
    }

    handleRequestPasswordReset = async(e, requestReset) => {
        e.preventDefault();
        await requestReset();
        this.setState({
            email: ''
        });
    }

    handlePasswordReset = async(e, resetPassword) => {
        e.preventDefault();
        await resetPassword();
        this.setState({
            password: '',
            confirmPassword: '',
        });
    }

    renderRequestResetPasswordForm() {
        return (
            <Mutation mutation={REQUEST_PASSWORD_RESET_MUTATION} variables={{ email: this.state.email }} >
                {(requestReset, payload) => (
                    <Form method="post" onSubmit={e => { this.handleRequestPasswordReset(e, requestReset); }}>
                        <fieldset disabled={payload.loading} aria-busy={payload.loading}>
                            <h2>Reset your Account&rsquo;s Password</h2>
                            {payload.error ? <ErrorMessage error={payload.error} /> : null}
                            {!payload.error && !payload.loading && payload.called && payload.data ? <ConfirmationMessage message={payload.data.requestPasswordReset.message} /> : null}
                            <label htmlFor="email">
                                Email
                                <input
                                    type="email"
                                    name="email"
                                    placeholder="email"
                                    value={this.state.email}
                                    onChange={this.saveToState}
                                />
                            </label>
                            <button type="submit">Reset</button>
                        </fieldset>
                    </Form>
                )}
            </Mutation>
        );
    }

    renderResetPasswordForm() {
        return (
            <Mutation
                mutation={RESET_PASSWORD_MUTATION}
                variables={{
                    resetToken: this.props.token,
                    password: this.state.password,
                    confirmPassword: this.state.confirmPassword
                }}
                refetchQueries={[{ query: CURRENT_USER_QUERY }]}>
                {(resetPassword, payload) => (
                    <Form method="post" onSubmit={e => { this.handlePasswordReset(e, resetPassword); }}>
                        <fieldset disabled={payload.loading} aria-busy={payload.loading}>
                            <h2>Reset your Account&rsquo;s Password</h2>
                            {payload.error ? <ErrorMessage error={payload.error} /> : null}
                            {!payload.error && payload.data ? <ConfirmationMessage message={`Password sucessfully reset for user ${payload.data.resetPassword.name}`} /> : null}
                            <label htmlFor="password">
                                Password:
                                <input
                                    type="password"
                                    name="password"
                                    placeholder="New password"
                                    value={this.state.password}
                                    onChange={this.saveToState}
                                />
                            </label>
                            <label htmlFor="confirmPassword">
                                Confirm the new password:
                                <input
                                    type="password"
                                    name="confirmPassword"
                                    placeholder="Confirm your password"
                                    value={this.state.confirmPassword}
                                    onChange={this.saveToState}
                                />
                            </label>
                            <button type="submit">Confirm</button>
                        </fieldset>
                    </Form>
                )}
            </Mutation>
        );
    }

    render() {
        return this.props.token ? this.renderResetPasswordForm() : this.renderRequestResetPasswordForm();
    }
}

ResetPassword.propTypes = {
    email: PropTypes.string,
    token: PropTypes.string,
};

export default ResetPassword;
