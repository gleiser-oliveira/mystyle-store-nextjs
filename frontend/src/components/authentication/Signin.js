import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import Link from 'next/link';
import { Form } from '../theme/generalStyles';
import { SIGNIN_MUTATION, CURRENT_USER_QUERY } from './AuthenticationOperationsGQL';
import { ErrorMessage, ConfirmationMessage } from '../message/Message';

class Signin extends Component {
    state = {
        name: '',
        email: '',
        password: '',
    };

    saveToState = (e) => {
        this.setState({ [e.target.name]: e.target.value });
    }

    handleSubmit = async(e, signin) => {
        e.preventDefault();
        const res = await signin();
        this.setState({
            name: '',
            email: '',
            password: '',
            message: `User ${res.data.signin.name} logged successfully!`,
        });
    }

    render() {
        return (
            <Mutation mutation={SIGNIN_MUTATION} variables={{ email: this.state.email, password: this.state.password }} refetchQueries={[{ query: CURRENT_USER_QUERY }]}>
                {(signin, payload) => (
                    <Form method='post' onSubmit={e => { this.handleSubmit(e, signin); }}>
                        <fieldset disabled={payload.loading} aria-busy={payload.loading}>
                            <h2>Sign into your Account</h2>
                            {payload.error ? <ErrorMessage error={payload.error} /> : null}
                            {!payload.error && this.state.message ? <ConfirmationMessage message={this.state.message} /> : null}
                            <label htmlFor='email'>
                                Email
                                <input
                                    type='email'
                                    name='email'
                                    placeholder='email'
                                    value={this.state.email}
                                    onChange={this.saveToState}
                                />
                            </label>
                            <label htmlFor='password'>
                                Password
                                <input
                                    type='password'
                                    name='password'
                                    placeholder='password'
                                    value={this.state.password}
                                    onChange={this.saveToState}
                                />
                            </label>
                            <Link href={{
                                pathname: '/resetPassword',
                                query: { email: this.state.email }
                            }}>
                                <a>Forgot your password? Click here!</a>
                            </Link>
                            <button type='submit'>Sign In!</button>
                        </fieldset>
                    </Form>
                )}
            </Mutation>
        );
    }
}

export default Signin;
