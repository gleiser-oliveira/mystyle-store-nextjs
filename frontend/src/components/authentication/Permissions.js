import React, { Component } from 'react';
import PropTypes from 'prop-types';
import NeedsPermission from './NeedsPermission';
import { Query, Mutation } from 'react-apollo';
import { ALL_PERMISSIONS_QUERY, ALL_USERS_QUERY, USERS_PAGINATION_QUERY, UPDATE_USER_PERMISSIONS_MUTATION } from './AuthenticationOperationsGQL';
import { ErrorMessage } from '../message/Message';
import Pagination from '../pagination/Pagination';
import { Center } from '../theme/generalStyles';
import { PermissionsGrid, AdminView } from './styles';
import { perPage } from '../../config';

class UserPermissions extends Component {
    state = {
        id: this.props.user.id,
        permissions: this.props.user.permissions,
    }

    handlePermissionChange = (e) => {
        let updatedPermissions = [...this.state.permissions];
        if (e.target.checked)
            updatedPermissions.push(e.target.value);
        else
            updatedPermissions = updatedPermissions.filter(permission => (permission !== e.target.value));

        this.setState({ permissions: updatedPermissions });
    }

    updatePermissionAction = (e, updatePermissions) => {
        updatePermissions();
    }

    render() {
        const permissionsInputs = this.props.allPermissions.map(permObj => (<div key={`${this.props.user.id}-permission-${permObj.permission}`} className="perm">
            <label htmlFor={`${this.props.user.id}-permission-${permObj.permission}`}>
                <input id={`${this.props.user.id}-permission-${permObj.permission}`} type="checkbox" value={permObj.permission} checked={this.state.permissions.includes(permObj.permission)} onChange={this.handlePermissionChange} />
            </label>
        </div>));

        const permissionsRow = <Mutation mutation={UPDATE_USER_PERMISSIONS_MUTATION} variables={{ id: this.state.id, permissions: this.state.permissions }}>
            {(updatePermissions, payload) => (
                [permissionsInputs, <div key={this.props.user.id} className='perm'><button onClick={(e) => { this.updatePermissionAction(e, updatePermissions); }}>UPDAT{payload.loading ? 'ING' : 'E'}</button></div>]
            )}
        </Mutation>;

        return permissionsRow;
    }
}

class Permissions extends Component {
    state = {
        pagLoading: false,
    }

    setLoading = (pag, isLoading) => {
        if (this.state[pag] !== isLoading)
            this.setState({ [pag]: isLoading });
    };

    renderPermissionsGrid = (permissionsData, usersData) => {
        return <Center>
            <AdminView>
                <PermissionsGrid>
                    <div className="col-u"></div>
                    <div className="col-p">
                        <div className="perms">
                            {permissionsData.map(perm => (<div key={perm.id} className="perm">{perm.description}</div>))}
                            <div className="perm">
                                <p>UPDATE</p>
                            </div>
                        </div>
                    </div>
                    <div className="col col-u">
                        {usersData.map(user => (<div key={user.id} className="user">{user.name}</div>))}
                    </div>
                    <div className="col col-p">
                        {usersData.map(user => (<div key={user.id} className="perms"><UserPermissions key={user.id} allPermissions={permissionsData} user={user} /></div>))}
                    </div>
                </PermissionsGrid>
            </AdminView>
            <Pagination
                page={this.props.page}
                loading={{ event: this.setLoading, name: 'pagLoading' }}
                query={USERS_PAGINATION_QUERY}
                nameObjects='Users'
                pathname='permissions'
                connectionName='usersConnection'
            />
        </Center>;
    }

    render() {
        return (
            <NeedsPermission neededPermission={['ADMIN']} user={this.props.userData} showMessage={true} >
                <Query query={ALL_PERMISSIONS_QUERY}>
                    {({ data, loading, error }) => {
                        if (loading)
                            return <p>Loading permissions...</p>;
                        if (error)
                            return <ErrorMessage error={error} />;
                        if (data)
                            return (
                                <Query query={ALL_USERS_QUERY} variables={{
                                    skip: this.props.page * perPage - perPage,
                                    first: perPage
                                }}>
                                    {payload => {
                                        if (payload.loading)
                                            return <p>Loading users...</p>;
                                        if (payload.data)
                                            return this.renderPermissionsGrid(data.permissions, payload.data.users);
                                    }}
                                </Query>
                            );
                    }}
                </Query>
            </NeedsPermission>
        );
    }
}

UserPermissions.propTypes = {
    user: PropTypes.shape({
        id: PropTypes.string,
        permissions: PropTypes.array,
    }),
    allPermissions: PropTypes.array,
};

Permissions.propTypes = {
    userData: PropTypes.number,
    page: PropTypes.number,
};

export default Permissions;
