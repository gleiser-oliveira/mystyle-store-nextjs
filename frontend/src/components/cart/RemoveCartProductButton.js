import React from 'react';
import PropTypes from 'prop-types';
import { Mutation } from 'react-apollo';
import { REMOVE_CART_PRODUCT_MUTATION } from './CartOperationsGQL';
import { CURRENT_USER_QUERY } from '../authentication/AuthenticationOperationsGQL';
import { RemoveButton } from './styles';

const RemoveCartProductButton = props => {
    return (
        <Mutation mutation={ REMOVE_CART_PRODUCT_MUTATION } variables={ { id: props.cartProductId } } refetchQueries={ [{ query: CURRENT_USER_QUERY }] }>
            {(removeCartProduct, payload) => (
                <RemoveButton disabled={payload.loading} onClick={removeCartProduct}>
                    &times;
                </RemoveButton>
            )}
        </Mutation>
    );
};

RemoveCartProductButton.propTypes = {
    cartProductId: PropTypes.string,
};

export default RemoveCartProductButton;
