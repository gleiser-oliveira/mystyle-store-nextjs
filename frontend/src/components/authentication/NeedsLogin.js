import React from 'react';
import PropTypes from 'prop-types';
import User from '../authentication/User';
import Signin from '../authentication/Signin';

const NeedsLogin = props => (
    <User>
        {payload => (
            payload.data.me ? props.children : <div>
                <p>You need to sign in to access this resource.</p>
                <Signin/>
            </div>
        )}
    </User>
);

NeedsLogin.propTypes = {
    children: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.object,
    ]),
};

export default NeedsLogin;
