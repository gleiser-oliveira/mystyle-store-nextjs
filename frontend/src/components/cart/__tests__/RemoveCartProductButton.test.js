import { mount } from 'enzyme';
import { MockedProvider } from '@apollo/react-testing';
import { ApolloConsumer } from 'react-apollo';
import toJSON from 'enzyme-to-json';
import RemoveCartProductButton from '../RemoveCartProductButton';
import { REMOVE_CART_PRODUCT_MUTATION } from '../CartOperationsGQL';
import { CURRENT_USER_QUERY } from '../../authentication/AuthenticationOperationsGQL';
import { updateWrapper, fakeUser, fakeCartProduct } from '../../../lib/testUtils';

const cartProd = fakeCartProduct();
const mocks = [
    {
        request: {
            query: CURRENT_USER_QUERY
        },
        result: {
            data: {
                me: {
                    ...fakeUser(),
                    cart: [cartProd],
                },
            },
        },
    },
    {
        request: {
            query: REMOVE_CART_PRODUCT_MUTATION,
            variables: { id: cartProd.id }
        },
        result: {
            data: {
                removeCartProduct: {
                    __typename: 'CartProduct',
                    id: cartProd.id,
                },
            },
        },
    },
    {
        request: {
            query: CURRENT_USER_QUERY
        },
        result: {
            data: {
                me: {
                    ...fakeUser(),
                    cart: [{ ...cartProd, quantity: cartProd.quantity - 1 }],
                },
            },
        },
    },
];

describe('testing the <RemoveCartProductButton/> component', () => {
    it('renders', async() => {
        const wrapper = mount(
            <MockedProvider>
                <RemoveCartProductButton cartProductId={cartProd.id}/>
            </MockedProvider>
        );

        expect(toJSON(wrapper.find('RemoveCartProductButton'))).toMatchSnapshot();
    });
    it('removes the product from the cart', async() => {
        let apolloClient;
        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <ApolloConsumer>
                    {client => {
                        apolloClient = client;
                        return <RemoveCartProductButton cartProductId={cartProd.id}/>;
                    }}
                </ApolloConsumer>
            </MockedProvider>
        );

        const { data: { me } } = await apolloClient.query({ query: CURRENT_USER_QUERY });
        await updateWrapper(wrapper);
        expect(me.cart).toHaveLength(1);
        expect(me.cart[0].quantity).toBe(2);
        expect(me.cart[0].product.price).toBe(5000);
        wrapper.find('button').simulate('click');
        await updateWrapper(wrapper, 50);
        const { data: { me: me2 } } = await apolloClient.query({ query: CURRENT_USER_QUERY });

        expect(me2.cart).toHaveLength(1);
        expect(me2.cart[0].quantity).toBe(1);
    });
});
