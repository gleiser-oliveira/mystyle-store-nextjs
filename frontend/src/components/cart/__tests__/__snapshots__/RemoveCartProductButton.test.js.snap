// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing the <RemoveCartProductButton/> component renders 1`] = `
<RemoveCartProductButton
  cartProductId="prod123"
>
  <Mutation
    mutation={
      Object {
        "definitions": Array [
          Object {
            "directives": Array [],
            "kind": "OperationDefinition",
            "name": Object {
              "kind": "Name",
              "value": "REMOVE_CART_PRODUCT_MUTATION",
            },
            "operation": "mutation",
            "selectionSet": Object {
              "kind": "SelectionSet",
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [
                    Object {
                      "kind": "Argument",
                      "name": Object {
                        "kind": "Name",
                        "value": "id",
                      },
                      "value": Object {
                        "kind": "Variable",
                        "name": Object {
                          "kind": "Name",
                          "value": "id",
                        },
                      },
                    },
                  ],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "removeCartProduct",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "id",
                        },
                        "selectionSet": undefined,
                      },
                    ],
                  },
                },
              ],
            },
            "variableDefinitions": Array [
              Object {
                "defaultValue": undefined,
                "directives": Array [],
                "kind": "VariableDefinition",
                "type": Object {
                  "kind": "NonNullType",
                  "type": Object {
                    "kind": "NamedType",
                    "name": Object {
                      "kind": "Name",
                      "value": "ID",
                    },
                  },
                },
                "variable": Object {
                  "kind": "Variable",
                  "name": Object {
                    "kind": "Name",
                    "value": "id",
                  },
                },
              },
            ],
          },
        ],
        "kind": "Document",
        "loc": Object {
          "end": 123,
          "start": 0,
        },
      }
    }
    refetchQueries={
      Array [
        Object {
          "query": Object {
            "definitions": Array [
              Object {
                "directives": Array [],
                "kind": "OperationDefinition",
                "name": Object {
                  "kind": "Name",
                  "value": "CURRENT_USER_QUERY",
                },
                "operation": "query",
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "me",
                      },
                      "selectionSet": Object {
                        "kind": "SelectionSet",
                        "selections": Array [
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "id",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "email",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "name",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "permissions",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "cart",
                            },
                            "selectionSet": Object {
                              "kind": "SelectionSet",
                              "selections": Array [
                                Object {
                                  "alias": undefined,
                                  "arguments": Array [],
                                  "directives": Array [],
                                  "kind": "Field",
                                  "name": Object {
                                    "kind": "Name",
                                    "value": "id",
                                  },
                                  "selectionSet": undefined,
                                },
                                Object {
                                  "alias": undefined,
                                  "arguments": Array [],
                                  "directives": Array [],
                                  "kind": "Field",
                                  "name": Object {
                                    "kind": "Name",
                                    "value": "quantity",
                                  },
                                  "selectionSet": undefined,
                                },
                                Object {
                                  "alias": undefined,
                                  "arguments": Array [],
                                  "directives": Array [],
                                  "kind": "Field",
                                  "name": Object {
                                    "kind": "Name",
                                    "value": "product",
                                  },
                                  "selectionSet": Object {
                                    "kind": "SelectionSet",
                                    "selections": Array [
                                      Object {
                                        "alias": undefined,
                                        "arguments": Array [],
                                        "directives": Array [],
                                        "kind": "Field",
                                        "name": Object {
                                          "kind": "Name",
                                          "value": "id",
                                        },
                                        "selectionSet": undefined,
                                      },
                                      Object {
                                        "alias": undefined,
                                        "arguments": Array [],
                                        "directives": Array [],
                                        "kind": "Field",
                                        "name": Object {
                                          "kind": "Name",
                                          "value": "price",
                                        },
                                        "selectionSet": undefined,
                                      },
                                      Object {
                                        "alias": undefined,
                                        "arguments": Array [],
                                        "directives": Array [],
                                        "kind": "Field",
                                        "name": Object {
                                          "kind": "Name",
                                          "value": "title",
                                        },
                                        "selectionSet": undefined,
                                      },
                                      Object {
                                        "alias": undefined,
                                        "arguments": Array [],
                                        "directives": Array [],
                                        "kind": "Field",
                                        "name": Object {
                                          "kind": "Name",
                                          "value": "image",
                                        },
                                        "selectionSet": undefined,
                                      },
                                      Object {
                                        "alias": undefined,
                                        "arguments": Array [],
                                        "directives": Array [],
                                        "kind": "Field",
                                        "name": Object {
                                          "kind": "Name",
                                          "value": "description",
                                        },
                                        "selectionSet": undefined,
                                      },
                                    ],
                                  },
                                },
                              ],
                            },
                          },
                        ],
                      },
                    },
                  ],
                },
                "variableDefinitions": Array [],
              },
            ],
            "kind": "Document",
            "loc": Object {
              "end": 390,
              "start": 0,
            },
          },
        },
      ]
    }
    variables={
      Object {
        "id": "prod123",
      }
    }
  >
    <styles__RemoveButton
      disabled={false}
      onClick={[Function]}
    >
      <button
        className="styles__RemoveButton-sc-1tfdw1p-4 kzpnub"
        disabled={false}
        onClick={[Function]}
      >
        ×
      </button>
    </styles__RemoveButton>
  </Mutation>
</RemoveCartProductButton>
`;
